### PR TITLE
[JAX] Support Ring Attention (Context Parallelism)

### PR DESCRIPTION
**Pull request by mingxu1067**
# Description

Add ring attention as an additional context parallel strategy to Jax fused attention API.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Adds a new context parallel strategy and unit tests.

# Checklist:

- [X] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [X] The functionality is complete
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 1]
```diff
@@ -1411,6 +1414,529 @@ def _cross_attn_bwd(
...
+@dataclass(frozen=True)
+class _FusedAttnCPWithP2PHelper:
+    """Helper class to assist with running the P2P ring strategy for CP attention."""
+
+    mesh: jax.sharding.Mesh
+    config: _FusedAttnConfig
+
+    @staticmethod
+    @cache
+    def use_scanloop():
```
Comment (mgoldfarb-nvidia) [inline transformer_engine/jax/cpp_extensions/attention.py:1426]: This is a bit ugly but a necessity of using scan loop which is preferred as it gives XLA more control over unrolling. We won't prevent not using scan but it currently warns the user to update their XLA flags. `NVTE_FUSED_RING_ATTENTION_USE_SCAN` if undefined will default to 1.

Hopefully `--xla_experimental_ignore_channel_id` becomes default in XLA:GPU at some point.

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 2]
```diff
@@ -1411,6 +1414,529 @@ def _cross_attn_bwd(
...
+            kv_seqlen,
+            q_seq_offsets,
+            k_seq_offsets,
+            seed,
+        ):
+            _not_used = jnp.zeros(0, dtype=v.dtype)
+
+            # Combine KV tensors if separate for better permute scheduling and performance.
+            # Eventually XLA should perform this automatically.
+            kv = helper.stack_kv(k, v)
```
Comment (mgoldfarb-nvidia) [inline transformer_engine/jax/cpp_extensions/attention.py:1569]: XLA:GPU does a poor job scheduling 2+ permutes with overlap so we manually combine K and V into our combined KV format. There is a plan to add functionality into XLA to do this fusion automatically.

**Review by zlsh80826 — COMMENTED** [review]
Thank you for the impressive work! I’ll need a few days to complete my review of the partition details within the ring attention.

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 3]
```diff
@@ -1411,6 +1414,529 @@ def _cross_attn_bwd(
...
+            dq = jnp.zeros_like(q)
+            dk_dv = helper.stack_kv(jnp.zeros_like(k), jnp.zeros_like(v))
+            dbias = jnp.zeros_like(bias)
+
+            def scan_kv_block(idx, carry):
+
+                kv, dq, dq_prev_step, dk_dv, dk_dv_prev_step, dbias, dbias_prev_step = carry
+
+                def accum_grads(dq, dq_prev_step, dk_dv, dk_dv_prev_step, dbias, dbias_prev_step):
+                    dq = dq + dq_prev_step
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/attention.py:1774]: I'm curious why we need to carry the per_step dq/dkv to the next iteration, can't we update the dq with dq_per_step in the current scan_kv_block?

File: transformer_engine/jax/cpp_extensions/misc.py [hunk 4]
```diff
@@ -167,3 +167,20 @@ def is_ffi_enabled():
...
+
+def get_xla_flag(flag: str, default=None):
+    """
+    Returns the value of an XLA_FLAGS environment variable if set or returns default.
+    """
+    xla_flags = []
+    if xla_flags_env := os.getenv("XLA_FLAGS"):
+        xla_flags.extend(xla_flags_env.split())
+    for flag_i in sorted(xla_flags):
+        if "=" in flag:
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/misc.py:180]: Thanks for catching- this code is a bit brittle and needs a unit test let me iterate on it.

File: transformer_engine/jax/cpp_extensions/misc.py [hunk 5]
```diff
@@ -167,3 +167,20 @@ def is_ffi_enabled():
...
+            name, val = flag_i.split("=", 2)
+            if name == flag:
+                return val
+        else:
+            return flag == flag_i  # bool flag with no value
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/misc.py]: I made a pass over this function and added some unit tests.

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 6]
```diff
@@ -1411,6 +1414,529 @@ def _cross_attn_bwd(
...
+    config: _FusedAttnConfig
+
+    @staticmethod
+    @cache
+    def use_scanloop():
+        """Returns true if the implementation will use a scan loop for iteration."""
+        use_scan = bool(int(os.getenv("NVTE_FUSED_RING_ATTENTION_USE_SCAN", "1")))
+        # nvbug(4675071): Disable the HLO verifier for channel ID checks.
+        # A WAR was added to XLA: https://github.com/openxla/xla/pull/16779
+        return use_scan and bool(get_xla_flag("--xla_experimental_ignore_channel_id", False))
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py]: I made this agnostic to the type and just consider `1` or `true` or no value to both be true. anything else is false which should cover any explicit case of setting that is accepted by XLA I think

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 7]
```diff
@@ -1411,6 +1414,503 @@ def _cross_attn_bwd(
...
+                    # This is for nested jax.lax.cond
+                    def jax_cond_wrap():
+                        if config.context_parallel_load_balanced:
+                            return lax.cond(
+                                (idx <= cp_rank), half_kv_no_mask_compute, half_q_no_mask_compute
+                            )
+                        return lax.cond((idx <= cp_rank), no_mask_compute, skip_compute)
+
+                    output_per_step, softmax_aux_per_step = lax.cond(
+                        idx == 0, causal_mask_compute, jax_cond_wrap
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/attention.py:1682]: Could you clarify why causal computation is applied to idx == 0? I initially thought that the causal mask would be needed for the trailing block instead.

File: transformer_engine/jax/cpp_extensions/attention.py [hunk 8]
```diff
@@ -1411,6 +1414,503 @@ def _cross_attn_bwd(
...
+            else:
+                for i in range(0, cp_size):
+                    carry = scan_kv_block(i, carry)
+            (kv, softmax_aux, output_per_steps, softmax_aux_per_steps) = carry
+
+            output = jnp.zeros(q.shape).astype(jnp.float32)
+            for idx in range(cp_size):
+                output = output + output_per_steps[idx].astype(jnp.float32) * jnp.exp(
+                    softmax_aux_per_steps[idx] - softmax_aux
+                ).transpose(0, 2, 1, 3)
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions/attention.py:1705]: Would it be feasible to transpose softmax_aux inside scan_block to allow for pipelining the transpose?

Comment (zlsh80826) [reply transformer_engine/jax/cpp_extensions/attention.py:1705]: If it doesn't affect the performance lots we can keep the transpose here

**Review by zlsh80826 — APPROVED** [review]
LGTM!

File: transformer_engine/jax/cpp_extensions/attention.py (hunk 7 above)
Comment (mgoldfarb-nvidia) [reply transformer_engine/jax/cpp_extensions/attention.py:1682]: Here is a picture of what happens in both unbalanced and load balanced case:

image

Each GPU_i starts with Q_i and KV_i at the start. These are the causal masked parts along the diagonal. All subsequence iterations are either skipped or non-masked partial computations. In the load balanced case this is where the "half KV" and "half Q" cases come from.
